turbine/src/broadcast_stage/broadcast_utils.rs

```diff
@@ -33,7 +48,23 @@ fn keep_coalescing_entries(
     // This slot is not over.
     last_tick_height < max_tick_height &&
     // We have not exceeded max batch byte count.
-    serialized_batch_byte_count < max_batch_byte_count
+    serialized_batch_byte_count < max_batch_byte_count &&
+    {
+        let bytes_to_fill_erasure_batch = data_shred_bytes_per_batch() - (serialized_batch_byte_count % data_shred_bytes_per_batch());
+        // We haven't tightly packed this batch.
+        (bytes_to_fill_erasure_batch as f64) > target_batch_pad_bytes()
+    }
+}
+
+// Dynamically determine the coalesce time based on the amount of entry data
+// that has been built up. If we only have a small amount, we can afford to wait
+// longer and avoid unnecessary padding. If we have a lot, we shouldn't wait as
+// long so we avoid delaying downstream validator replay.
+fn max_coalesce_time(serialized_batch_byte_count: u64, max_batch_byte_count: u64) -> Duration {
+    // Compute the fraction of the target batch that has been filled.
+    // Constrain to 75% to ensure we allow some time for coalescing.
+    let ratio = (serialized_batch_byte_count as f64 / max_batch_byte_count as f64).min(0.75);
```

Comment:
```suggestion
    let ratio = (serialized_batch_byte_count as f64 / max_batch_byte_count.max(1) as f64).min(0.75);
```
Just a bit of safety agains zero divide?

Reply:
adding an assert to verify this. It shouldn't ever happen because `max_batch_byte_count` get set to `next_full_batch_byte_count.max(target_batch_bytes_default());` and `target_batch_bytes_default()` is essentially a constant